📖  Update the Kubebuilder book tutorials for v3

hickeyma opened the pull request:
This PR updates the Kubebuilder book tutorials to align with the current version, v3. This fix will help users navigate the book with success.

Fixes #2273

hickeyma commented:
This is a WIP as I traverse the book. The initial docs (intro, quick start and arch) are reviewed. Next will be the cronjob tutorial.

hickeyma commented:
@camilamacedo86 First commit after running  https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/cronjob-tutorial/testdata/generate_cronjob.sh. Let me kjnow if going in the right direction.

camilamacedo86 commented on docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_types.go:
```diff
@@ -13,160 +13,36 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-As previously noted, we don't need to change this, except to mark that
-we want a status subresource, so that we behave like built-in kubernetes types.
-*/
-
```
You need to come back to the texts on top as well. 
I usually compare that with the IDEA and ensure that we just update what changed. 
So, you need to 
a) run the scripts to re-gen the scaffolds ( base ones )
b) run make in the dirs of the projects to generate the files
c) revert the files that was removed and are missing 
d) compare with master to add again the code on top + text and explanations.

hickeyma replied:
Thanks for the feedback @camilamacedo86. I have updated the commit. Some build issues to cleanup.

hickeyma commented:
Unit test failing. Just wondering if this is a glitch or not?

```
STEP: tearing down the test environment
Failure [20.007 seconds]
[AfterSuite] AfterSuite 
/home/runner/work/kubebuilder/kubebuilder/docs/book/src/cronjob-tutorial/testdata/project/controllers/suite_test.go:152

  Unexpected error:
      <errors.aggregate | len:1, cap:1>: [
          <*errors.errorString | 0xc0008499d0>{
              s: "timeout waiting for process kube-apiserver to stop",
          },
      ]
      timeout waiting for process kube-apiserver to stop
  occurred

  /home/runner/work/kubebuilder/kubebuilder/docs/book/src/cronjob-tutorial/testdata/project/controllers/suite_test.go:155
```

hickeyma commented:
/retest

camilamacedo86 commented:
HI @hickeyma, 

We need to solve : https://github.com/kubernetes-sigs/controller-runtime/issues/1571

camilamacedo86 commented on docs/book/src/cronjob-tutorial/testdata/project/controllers/suite_test.go:
```diff
@@ -127,22 +126,24 @@ var _ = BeforeSuite(func() {
 	*/
 
-	k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
+	//TODO: Fix this as it is disabled because following error when tearing down:
```
If we add the TODO here then that will appears the docs :-(

hickeyma replied:
Test is enabled again

hickeyma commented:
Thanks for the feedback @camilamacedo86. Blocked waiting on https://github.com/kubernetes-sigs/kubebuilder/pull/2325

hickeyma commented:
@camilamacedo86 I updated the failing tests and they are now passing. I have updated the controller-runtime `envTest` issue in https://github.com/kubernetes-sigs/controller-runtime/issues/1571#issuecomment-914401589. I had to change how the test environment is brought down when a controller is started. I have updated `cronjob_tutoria`l test in this PR accordingly.

estroz reviewed (approved):
The tutorial envtest bits will need to be changed when teardown starts working properly again. For now the comments are explanatory enough.

/lgtm
/approve

hickeyma commented:
Thanks for the reviews @camilamacedo86 @estroz, and for merging @estroz.

estroz commented:
Thanks for your contribution!
